### PR TITLE
[clangd] Support parsing comments without ASTContext

### PR DESCRIPTION
**Pull request by tom-anders**
This is in preparation for implementing doxygen parsing, see discussion in https://github.com/clangd/clangd/issues/529.

(Old Phabricator review: https://reviews.llvm.org/D143112)

### CODE REVIEW

**Comment by trustytrojan** [comment]
When will this ever get merged? Would really like to see `vscode-clangd` finally be able to render Doxygen comments and/or Markdown comments.

**Comment by aaronliu0130** [comment]
Well, the last ready-to-go PR was from September (#65448), so probably a while.

**Comment by trustytrojan** [comment]
@aaronliu0130 Do you know if a patch for this exists?

**Comment by aaronliu0130** [comment]
As mentioned in https://github.com/clangd/clangd/issues/529#issuecomment-1938730781, they're [here](https://aur.archlinux.org/cgit/aur.git/tree/hover-doxygen-trunk.patch?h=clangd-opt-git). Hopefully you can follow [the build instructions](https://github.com/llvm/llvm-project/tree/main/clang-tools-extra/clangd#readme).

**Comment by tfninjadoom** [comment]
I'm new to the processes of open-source software, so pardon my confusion. I just read through [this](https://github.com/clangd/clangd/issues/529) issue and am wondering: 
What's preventing this PR from being merged right now? All checks seem to have passed.

**Comment by aaronliu0130** [comment]
LLVM has 2.5k PRs to review while working on their own changes

**Comment by cor3ntin** [comment]
PRs need to be reviewed before they can be merged.
Feel free to ping PR once a week (at most) if no one gets to it.
This seem to have flown under the radar. sorry about that

**Comment by aaronliu0130** [comment]
What's "ping PR"? Do you mean pinging the LLVM organization?

**Comment by HighCommander4** [comment]
> What's "ping PR"? Do you mean pinging the LLVM organization?

It just means posting a comment on the PR that says "ping", as a reminder for reviewers.

(But please only do that if you're the patch author, or, in the case of a patch author who's not active any more, if you're volunteering to take over the patch from the patch author and do the work of addressing any remaining review comments.)

>> What's preventing this PR from being merged right now? All checks seem to have passed.
> PRs need to be reviewed before they can be merged.

Yes, code reviewer bandwidth is the bottleneck for clangd patches at the moment.

I have some other patches in my review queue which I've been trying to get to (and unfortunately haven't had much time to spend on them), so I'm hoping Kadir can review this one, especially as he has reviewed earlier versions of this and related patches and has more familiarity with the affected code areas.

But if no one else gets to this, I would like to at some point.

**Comment by db7** [comment]
> > What's "ping PR"? Do you mean pinging the LLVM organization?
> 
> It just means posting a comment on the PR that says "ping", as a reminder for reviewers.
> 
> (But please only do that if you're the patch author, or, in the case of a patch author who's not active any more, if you're volunteering to take over the patch from the patch author and do the work of addressing any remaining review comments.)
> 
> > > What's preventing this PR from being merged right now? All checks seem to have passed.
> > 
> > 
> > PRs need to be reviewed before they can be merged.
> 
> Yes, code reviewer bandwidth is the bottleneck for clangd patches at the moment.
> 
> I have some other patches in my review queue which I've been trying to get to (and unfortunately haven't had much time to spend on them), so I'm hoping Kadir can review this one, especially as he has reviewed earlier versions of this and related patches and has more familiarity with the affected code areas.
> 
> But if no one else gets to this, I would like to at some point.

If seems this PR is still stuck.

File: clang/include/clang/Basic/SourceManager.h [hunk 1]
```diff
@@ -1971,7 +1971,7 @@ class BeforeThanCompare<SourceRange> {
 };
 
 /// SourceManager and necessary dependencies (e.g. VFS, FileManager) for a
-/// single in-memorty file.
+/// single in-memory file.
```
Comment (erichkeane) [inline clang/include/clang/Basic/SourceManager.h:1974]: Unrelated chage, please put this in a different patch.

File: clang-tools-extra/clangd/CodeCompletionStrings.cpp [hunk 2]
```diff
@@ -316,5 +319,26 @@ std::string getReturnType(const CodeCompletionString &CCS) {
...
+comments::FullComment *parseComment(llvm::StringRef Comment,
+                                    llvm::BumpPtrAllocator &Allocator,
+                                    comments::CommandTraits &Traits) {
+  // The comment lexer expects markers, so add them back
+  auto CommentWithMarkers = "/*" + Comment.str() + "*/";
+
+  SourceManagerForFile SourceMgrForFile("mock_file.cpp", CommentWithMarkers);
+  SourceManager &SourceMgr = SourceMgrForFile.get();
+
+  comments::Lexer L(Allocator, SourceMgr.getDiagnostics(), Traits,
```
Comment (erichkeane) [inline clang-tools-extra/clangd/CodeCompletionStrings.cpp:331]: The order of these constructors seems odd.  I would expect us to build these up in the 'correct' order of dependence/execution/etc.  So I think you should swap the `Lexer` and `Sema` lines.

File: clang-tools-extra/clangd/CodeCompletionStrings.cpp [hunk 3]
```diff
@@ -316,5 +319,26 @@ std::string getReturnType(const CodeCompletionString &CCS) {
...
+                    SourceMgr.getLocForStartOfFile(SourceMgr.getMainFileID()),
+                    CommentWithMarkers.data(),
+                    CommentWithMarkers.data() + CommentWithMarkers.size());
+  comments::Sema S(Allocator, SourceMgr, SourceMgr.getDiagnostics(), Traits,
+                   nullptr);
```
Comment (erichkeane) [inline clang-tools-extra/clangd/CodeCompletionStrings.cpp:336]: we need a `/*stuff=*/` style comment for the `nullptr` here.

**Review by erichkeane — COMMENTED** [review]
I don't know much about the clang-d implementation, but this seems reasonable.

**Comment by tom-anders** [comment]
Hi everyone, I shifted focus to other projects and currently don't have time to work on this PR anymore. If somebody wants to take over, feel free to do so

**Comment by cor3ntin** [comment]
@AaronBallman @erichkeane what do you want to do here?

**Comment by AaronBallman** [comment]
> @AaronBallman @erichkeane what do you want to do here?

Because the author is no longer available for the time being, I'd recommend we close the PR. Someone can always come along to commandeer it if they have the bandwidth to work on it. WDYT?

**Comment by erichkeane** [comment]
> > @AaronBallman @erichkeane what do you want to do here?
> 
> Because the author is no longer available for the time being, I'd recommend we close the PR. Someone can always come along to commandeer it if they have the bandwidth to work on it. WDYT?

Closing due to author nonavailability.  Commandeering encouraged!
